presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -429,8 +432,30 @@ public ActualProperties visitJoin(JoinNode node, List<ActualProperties> inputPro
                             .unordered(true)
                             .build();
                 case FULL:
-                    // We can't say anything about the partitioning scheme because any partition of
-                    // a hash-partitioned join can produce nulls in case of a lack of matches
+                    if (probeProperties.getNodePartitioning().isPresent()) {
+                        Partitioning nodePartitioning = probeProperties.getNodePartitioning().get();
+                        ImmutableList.Builder<Partitioning.ArgumentBinding> outputArguments= ImmutableList.builder();
+                        for (Partitioning.ArgumentBinding argument : nodePartitioning.getArguments()) {
+                            if (argument.isVariable()) {
+                                Symbol column = argument.getColumn();
+                                for (JoinNode.EquiJoinClause equality : node.getCriteria()) {
+                                    if (equality.getLeft().equals(column) || equality.getRight().equals(column)) {
+                                        outputArguments.add(expressionBinding(new CoalesceExpression(ImmutableList.of(equality.getLeft().toSymbolReference(), equality.getRight().toSymbolReference()))));
+                                    }
+                                }
+                            }
+                            else if (argument.isConstant()) {
```

Comment:
This is the only new changes added. The previous implementation didn't handle constant in `ArgumentBinding`. As a result, if the input are partitioned on `(id, constant)`, the constant is dropped in the output, and we consider the output to be only partitioned on `(coalesce(id_left, id_right))`. This conclusion would be correct logically. However, since we use the partition columns to compute hash, and `hash(id) != hash(id, constant)`, this assumption is unsafe.

A better way to write this is to use the `ActualProperties.translate` API as used in other parts of this class (for example, in handle other types of JOIN above). However, this would require adding another `Function<Symbol, Optional<Expression>> translator`. This will make the overall API for translate really ugly, if not impossible. A better way is to convert Symbol to SymbolReference / Expression and combine all translators to a `Function<Expression, Optional<Expression>>` translator. I gave it a try, and it's a massive change because other classes, like `PreferredProperties` and `LocalProperty` also uses the same translator, and all related classes expect `Symbol`.

A even better way is to convert all of this to RowExpression, from PlanNode. Instead of `Symbol`, PlanNode should use `VariableReferenceExpression`. In that way, ArgumentBinding can combine `Symbol`, `Expression` and `NullableValue` under `RowExpression`. That's an even bigger refactoring. So I decided to put this fix up first. And get people's feedback on how should we proceed.

Another caveat is that, there's no easy way to test this from the plan test. Our plan tests does have inspect details in argument bindings. The difference of the query plan before and after this fix is in whether constant would be pushed to the source stage (if one side is partitioned on `(id, constant)`, the source stage of the other side will be partitioned on `(id, constant)` as well).

Happy to chat in person if this is still confusing.

Reply:
The fix on line 447 to 449 actually looks good to me, assuming this is the only change :).  I am still looking at other parts of the PR, digesting other's previous comments to get a comprehensive picture :). 

I find https://github.com/prestodb/presto/pull/12013#discussion_r241564826 useful to understand the relationship between `Partitioning` and `PartitioningHandle`. Think about `PartitioningHandle` something similar to `MethodHandle`, it knows how to evaluate the hash code, but don't know anything about the argument. And thus `Partitioning` provides the context about arguments. 

Some arguments are column/variables, so they are represented as `Symbol` (and we added `Expression` as well, in this PR). Another type of argument can be constant (on use case we noted recently is JOIN with constant filters: https://github.com/prestodb/presto/pull/12568#issuecomment-479106934).

In theory, if `PartitioningHandle` can do `bindTo` similar to `MethodHandle`, we can claim the `Partitioning` has only 1 argument binding over the new `PartioningHandle`.  But before that,  I see the pattern of unchanged `PartitioningHandle` + constant argument binding widely used in today's Presto codebase.

I do have a question:
- Before this code, in theory the generated `Partitioning` is *invalid*.  In your example,  `PartitioningHandle` asks for 2 argument, but the `ArgumentBinding` only provides 1... :). I understand today there is no way to check the number of arguments in `PartitioningHandle`. But the `Partitioning` shouldn't also be compatible with another *valid* `Partitioning` and thus we will always insert `Exchange`? 

Do you mind give me an example reproduce the issue so I can debug the code to understand better and where did I understand wrong? ;)